core/src/main/java/com/linecorp/armeria/common/logging/DefaultContentPreviewFactory.java

```diff
@@ -0,0 +1,107 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+
+import io.netty.buffer.ByteBuf;
+
+final class DefaultContentPreviewFactory implements ContentPreviewerFactory {
+
+    private static final List<String> textSubTypes = ImmutableList.of("json", "xml");
+    private static final List<String> textSubTypeSuffixes = ImmutableList.of("+json", "+xml");
+
+    private final int maxLength;
+    private final List<PreviewSpec> previewSpecs;
+    private final Charset defaultCharset;
+
+    DefaultContentPreviewFactory(List<PreviewSpec> previewSpecs, int maxLength, Charset defaultCharset) {
+        this.maxLength = maxLength;
+        this.previewSpecs = previewSpecs;
+        this.defaultCharset = defaultCharset;
+    }
+
+    @Override
+    public ContentPreviewer requestContentPreviewer(RequestContext ctx, RequestHeaders headers) {
+        return contentPreviewer(ctx, headers);
+    }
+
+    @Override
+    public ContentPreviewer responseContentPreviewer(RequestContext ctx, ResponseHeaders resHeaders) {
+        return contentPreviewer(ctx, resHeaders);
+    }
+
+    private ContentPreviewer contentPreviewer(RequestContext ctx, HttpHeaders headers) {
+        for (PreviewSpec previewSpec : previewSpecs) {
+            if (previewSpec.predicate().test(ctx, headers)) {
+                switch (previewSpec.mode()) {
+                    case TEXT:
+                        final Charset charset = charset(headers, previewSpec);
+                        return new TextContentPreviewer(maxLength, charset);
+                    case BINARY:
+                        final BiFunction<? super HttpHeaders, ? super ByteBuf, String> producer =
+                                previewSpec.producer();
+                        assert producer != null;
+                        return new ProducerBasedContentPreviewer(maxLength, headers, producer);
+                    case DISABLED:
+                        return ContentPreviewer.disabled();
+                }
+            }
+        }
+
+        final MediaType contentType = headers.contentType();
+        if (contentType != null) {
+            final Charset charset = contentType.charset();
+            if (charset != null) {
+                return new TextContentPreviewer(maxLength, charset);
+            }
+
+            if ("text".equals(contentType.type()) ||
+                textSubTypes.contains(contentType.subtype()) ||
+                textSubTypeSuffixes.stream().anyMatch(contentType.subtype()::endsWith) ||
+                contentType.is(MediaType.FORM_DATA)) {
```

Comment:
You don't need to address this preview. I just thought we can optimize this after releasing 1.0. 
`MediaType` class has `KNOWN_TYPES`. If the `contentType` is `KNOWN_TYPES` these predicate results are cachable with limited size. :-)

Reply:
Ah yes, it's probably a good idea to cache the result. But we have to check the performance because we have to do it in thread safely and the size of cached should be limited which means we have to use caffeine. 😉

Reply:
Yes, we should prove optimized code with benchmarks. By the way I thought we might build this cache when class is loading so we don’t care about thread safety.

Reply:
How could we do that? I don't probably understand what you mean. Could you elaborate more please? 😄

Reply:
Let me try this 😉and will give a feedback.

Reply:
I thought this kind of cache. :-) Even though I didn't run a benchmark yet. 
```diff
diff --git a/core/src/main/java/com/linecorp/armeria/common/MediaType.java b/core/src/main/java/com/linecorp/armeria/common/MediaType.java
index 9875a1ae0..e7e504290 100644
--- a/core/src/main/java/com/linecorp/armeria/common/MediaType.java
+++ b/core/src/main/java/com/linecorp/armeria/common/MediaType.java
@@ -127,6 +127,10 @@ public final class MediaType {
 
     private static final Map<MediaType, MediaType> KNOWN_TYPES = Maps.newHashMap();
 
+    public static Map<MediaType, MediaType> knownTypes() {
+        return KNOWN_TYPES;
+    }
+
     private static MediaType createConstant(String type, String subtype) {
         MediaType mediaType =
                 addKnownType(new MediaType(type, subtype, ImmutableListMultimap.of()));
diff --git a/core/src/main/java/com/linecorp/armeria/common/logging/DefaultContentPreviewFactory.java b/core/src/main/java/com/linecorp/armeria/common/logging/DefaultContentPreviewFactory.java
index 6a4f2d665..405f69b45 100644
--- a/core/src/main/java/com/linecorp/armeria/common/logging/DefaultContentPreviewFactory.java
+++ b/core/src/main/java/com/linecorp/armeria/common/logging/DefaultContentPreviewFactory.java
@@ -16,8 +16,12 @@
 
 package com.linecorp.armeria.common.logging;
 
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
 import java.nio.charset.Charset;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.function.BiFunction;
 
 import com.google.common.collect.ImmutableList;
@@ -35,6 +39,12 @@ final class DefaultContentPreviewFactory implements ContentPreviewerFactory {
     private static final List<String> textSubTypes = ImmutableList.of("json", "xml");
     private static final List<String> textSubTypeSuffixes = ImmutableList.of("+json", "+xml");
 
+    private static final Map<MediaType, Boolean> previewableTextMediaTypes =
+            MediaType.knownTypes()
+                     .entrySet()
+                     .stream()
+                     .collect(toImmutableMap(Entry::getKey, entry -> isTextContent0(entry.getValue())));
+
     private final int maxLength;
     private final List<PreviewSpec> previewSpecs;
     private final Charset defaultCharset;
@@ -55,6 +65,22 @@ final class DefaultContentPreviewFactory implements ContentPreviewerFactory {
         return contentPreviewer(ctx, resHeaders);
     }
 
+    private static boolean isTextType(MediaType contentType) {
+        final Boolean isTextType = previewableTextMediaTypes.get(contentType);
+        if (isTextType != null) {
+            return isTextType;
+        } else {
+            return isTextContent0(contentType);
+        }
+    }
+
+    private static boolean isTextContent0(MediaType contentType) {
+        return "text".equals(contentType.type()) ||
+               textSubTypes.contains(contentType.subtype()) ||
+               textSubTypeSuffixes.stream().anyMatch(contentType.subtype()::endsWith) ||
+               contentType.is(MediaType.FORM_DATA);
+    }
+
     private ContentPreviewer contentPreviewer(RequestContext ctx, HttpHeaders headers) {
         for (PreviewSpec previewSpec : previewSpecs) {
             if (previewSpec.predicate().test(ctx, headers)) {
@@ -80,10 +106,7 @@ final class DefaultContentPreviewFactory implements ContentPreviewerFactory {
                 return new TextContentPreviewer(maxLength, charset);
             }
 
-            if ("text".equals(contentType.type()) ||
-                textSubTypes.contains(contentType.subtype()) ||
-                textSubTypeSuffixes.stream().anyMatch(contentType.subtype()::endsWith) ||
-                contentType.is(MediaType.FORM_DATA)) {
+            if (isTextType(contentType)) {
                 return new TextContentPreviewer(maxLength, defaultCharset);
             }
         }
```

Reply:
Ah, I was thinking caching every media type of request, but yes this is a good approach and we don't even need the benchmark because it's obviously winning.
As I left the comment, I will look into this later with other logic which handles text content type. Thanks!